jsrc/conversions.cpp

```diff
@@ -0,0 +1,637 @@
+        (void)value;
+        return true;
+    } else {
+        return std::numeric_limits<T>::lowest() <= value && value <= std::numeric_limits<T>::max();
+    }
+}
+
+template <typename T, typename V>
+[[nodiscard]] constexpr auto
+in_range() -> bool {
+    return in_range<T>(std::numeric_limits<V>::lowest()) && in_range<T>(std::numeric_limits<V>::max());
+}
+
+template <typename T>
+struct is_optional : std::false_type {};
+
+template <typename T>
+struct is_optional<std::optional<T>> : std::true_type {};
+
+template <typename T>
+[[nodiscard]] auto
+value_if(bool cond, T value) -> std::optional<T> {
+    return cond ? std::optional(value) : std::nullopt;
+}
+
+template <typename From, typename To, typename Transform>
+[[nodiscard]] auto
+convert(J jt, array w, void *yv, Transform t) -> bool {
+    auto *v = pointer_to_values<From>(w);
+    auto *result = static_cast<To *>(yv);
+    if constexpr (is_optional<decltype(t(*v))>::value) {
+        for (int64_t i = 0; i < AN(w); ++i) {
+            auto opt = t(v[i]);
+            if (!opt) return false;
+            result[i] = opt.value();
+        }
+    } else {
+        std::transform(v, v + AN(w), result, t);
+    }
+    return true;
+}
+
+template <typename From, typename To>
+[[nodiscard]] auto
+convert(J jt, array w, void *yv) -> bool {
+    if constexpr (std::is_same_v<To, Z>) {
+        return convert<From, To>(jt, w, yv, [](auto v) { return To{.re = static_cast<D>(v), .im = {}}; });
+    } else if constexpr (!in_range<To, From>()) {
+        return convert<From, To>(jt, w, yv, [](auto v) { return value_if(in_range<To>(v), v); });
+    } else {
+        (void)jt;
+        auto *v = pointer_to_values<From>(w);
+        std::copy(v, v + AN(w), static_cast<To *>(yv));
+        return true;
+    }
+}
+
+template <>
+[[nodiscard]] auto
+convert<D, bool>(J jt, array w, void *yv, D fuzz) -> bool {
+    auto const infinity = [](auto p) { return p < -2 || 2 < p; };
+    return convert<D, bool>(jt, w, yv, [&](auto p) {
+        return value_if(!infinity(p) && (p == 0.0 || fuzzy_equal(p, 1.0, fuzz)), p != 0.0);
+    });
+}
+
+template <>
+[[nodiscard]] auto
+convert<D, I>(J jt, array w, void *yv, D fuzz) -> bool {
+    return convert<D, I>(jt, w, yv, [&](auto p) -> std::optional<I> {
+        auto const q = jround(p);
+        if (!(p == q || fuzzy_equal(p, q, fuzz))) {
+            return std::nullopt;  // must equal int, possibly out of range
+        }
+        // out-of-range values don't convert, handle separately
+        if (p < static_cast<D> IMIN) {
+            return value_if(p >= IMIN * (1 + fuzz), IMIN);
+        }  // if tolerantly < IMIN, error; else take IMIN
+        else if (p >= FLIMAX) {
+            return value_if(p <= -static_cast<D> IMIN * (1 + fuzz), IMAX);
+        }  // if tolerantly > IMAX, error; else take IMAX
+        return q;
+    });
+}
+
+template <>
+[[nodiscard]] auto
+convert<Z, D>(J jt, array w, void *yv, D fuzz) -> bool {
+    if (fuzz != 0.0) {
```

Comment:
Is this ok?

Reply:
Original test was `if (fuzz)`, not sure if these have differences